scapy/contrib/pnio_rtc.py

```diff
@@ -0,0 +1,511 @@
+                         length_from=lambda p: p[PNIORealTimeRawData].length()
+                         # pylint: enable=undefined-variable
+                        ),
+        ]
+
+    def __init__(self, _pkt="", post_transform=None, _internal=0,
+                 _underlayer=None, config=None, **fields):
+        """
+        length=None means that the length must be managed by the user. If it's
+        defined, the field will always be length-long (padded with '\\x00' if
+        needed)
+        """
+        self._config = config
+        Packet.__init__(self, _pkt=_pkt, post_transform=post_transform,
+                        _internal=_internal, _underlayer=_underlayer, **fields)
+
+    def copy(self):
+        pkt = Packet.copy(self)
+        # pylint: disable=protected-access
+        pkt._config = self._config
+        # pylint: enable=protected-access
+        return pkt
+
+    def clone_with(self, *args, **kargs):
+        pkt = Packet.clone_with(self, *args, **kargs)
+        # pylint: disable=protected-access
+        pkt._config = self._config
+        # pylint: enable=protected-access
+        return pkt
+
+    def length(self):
+        """Get the length of the raw data"""
+        # Manage the length of the packet if a length is provided
+        return  self._config['length']
+
+# Make sure an IOPS follows a data
+bind_layers(PNIORealTimeRawData, PNIORealTimeIOxS)
+
+
+
+###############################
+## PROFINET Real-Time Fields ##
+###############################
+
+class LowerLayerBoundPacketListField(PacketListField):
+    """PacketList which binds each underlayer of packets to the current pkt"""
+    def m2i(self, pkt, m):
+        return self.cls(m, _underlayer=pkt)
+
+class NotionalLenField(Field):
+    """A len fields which isn't present in the machine representation, but is
+    computed from a given lambda"""
+    __slots__ = ['length_from', 'count_from']
+    def __init__(self, name, default, length_from=None, count_from=None):
+        Field.__init__(self, name, default)
+        self.length_from = length_from
+        self.count_from = count_from
+
+    def addfield(self, pkt, s, val):
+        return s   # Isn't present in the machine packet
+
+    def getfield(self, pkt, s):
+        val = None
+        if self.length_from is not None:
+            val = self.length_from(pkt, s)
+        elif self.count_from is not None:
+            val = self.count_from(pkt, s)
+        return s, val
+
+
+###############################
+## PNIORealTime Configuration #
+###############################
+
+# PNRT_CONFIG is a dict which contains data layout for each Ethernet
+# communications. It must be formatted as such:
+# {(Ether.src, Ether.dst): [(start, type, config), ...]}
+# start: index of a data field from the END of the data buffer (-1, -2, ...)
+# type: class to be instanciated to represent these data
+# config: a config dict, given to the type class constructor
+PNIO_RTC_CONFIG = {}
```

Comment:
I wonder if this could be stored in the `conf` object I guess that would be more convenient for end users.

You could define `conf.contribs` as a `dict()` and, for example, store your configuration variables in `conf.contribs["PNIO"]`.